internal/appsec/remoteconfig.go

```diff
@@ -70,29 +70,49 @@ updateLoop:
 			statuses = mergeMaps(statuses, status)
 			r.AddEdit("asmdata", config.RulesFragment{RulesData: rulesData})
 		case rc.ProductASMDD:
-			// Switch the base rules of the RulesManager if the config received through ASM_DD is valid
-			// If the config was removed, switch back to the static recommended rules
-			if len(u) > 1 { // Don't process configs if more than one is received for ASM_DD
-				log.Debug("appsec: Remote config: more than one config received for ASM_DD. Updates won't be applied")
-				err = errors.New("more than one config received for ASM_DD")
-				statuses = mergeMaps(statuses, statusesFromUpdate(u, true, err))
-				break updateLoop
-			}
+			removalFound := false
+			var newBasePath string
+			var newBaseData []byte
 			for path, data := range u {
+				// Already seen a removal or an update, return an error
+				if (data == nil && removalFound) || (data != nil && newBaseData != nil) {
+					// Multiple real updates in the map, return an error
+					err = errors.New("more than one config received for ASM_DD")
+					statuses = mergeMaps(statuses, statusesFromUpdate(u, true, err))
+					break updateLoop
+				}
+
 				if data == nil {
+					removalFound = true
+					continue
+				}
+
+				// Save the new base path and data and only make the update after cycle through all received configs
+				// This makes sure that we don't update the ruleset in case the update is invalid (ex: several non nil configs)
+				newBasePath = path
+				newBaseData = data
+			}
+			// update with data = nil means the config was removed, so we switch back to the default rules
+			// only happens if no update was found, otherwise it could revert the switch to the new base rules
+			if newBaseData == nil {
+				if removalFound {
 					log.Debug("appsec: Remote config: ASM_DD config removed. Switching back to default rules")
 					r.ChangeBase(config.DefaultRulesFragment(), "")
-					break
-				}
-				var newBase config.RulesFragment
-				if err = json.Unmarshal(data, &newBase); err != nil {
-					log.Debug("appsec: Remote config: could not unmarshall ASM_DD rules: %v", err)
-					statuses[path] = genApplyStatus(true, err)
-					break updateLoop
+					statuses = mergeMaps(statuses, statusesFromUpdate(u, true, nil))
 				}
-				log.Debug("appsec: Remote config: switching to %s as the base rules file", path)
-				r.ChangeBase(newBase, path)
+				continue
+			}
+
+			// Switch the base rules of the RulesManager if the config received through ASM_DD is valid
+			var newBase config.RulesFragment
+			if err := json.Unmarshal(newBaseData, &newBase); err != nil {
+				log.Debug("appsec: Remote config: could not unmarshall ASM_DD rules: %v", err)
+				statuses[newBasePath] = genApplyStatus(true, err)
+				break updateLoop
```

Comment:
![image](https://github.com/DataDog/dd-trace-go/assets/411689/9437af65-53cb-4aa8-bc88-b9aa3fd87ef6)

And there I was thinking this code looked indented a little deep... 🫣

Optional, but it MIGHT improve readability if some slices of this function were refactored out (but it might also further obscure the point so... not sure it's a good idea).

Reply:
I actually tried to do it but it led me nowhere so I just reverted it. I agree that we can round corners a bit but that's all unfortunately. Maybe @Hellzy found a betterway ™️  but I did not

Reply:
Yeah, nothing too convincing on my end either for now. Let's merge it like this and follow-up on this later on.